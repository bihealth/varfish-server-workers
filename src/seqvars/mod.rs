pub mod aggregate;
pub mod ingest;
pub mod prefilter;
pub mod query;
