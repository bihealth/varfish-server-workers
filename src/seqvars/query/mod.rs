//! Code implementing the "seqvars query" sub command.

mod schema;

use std::time::Instant;

use clap::{command, Parser};

use rand_core::SeedableRng;

use crate::{common::trace_rss_now, common::GenomeRelease};

/// Command line arguments for `seqvars query` sub command.
#[derive(Parser, Debug)]
#[command(author, version, about = "Run query for seqvars", long_about = None)]
pub struct Args {
    /// Genome release to assume.
    #[arg(long, value_enum)]
    pub genome_release: GenomeRelease,
    /// Path to worker database to use for querying.
    #[arg(long, required = true)]
    pub path_db: String,
    /// Path to query JSON file.
    #[arg(long, required = true)]
    pub path_query_json: String,
    /// Path to input TSV file.
    #[arg(long, required = true)]
    pub path_input: String,
    /// Path to the output TSV file.
    #[arg(long, required = true)]
    pub path_output: String,

    /// Optional maximal number of total records to write out.
    #[arg(long)]
    pub max_results: Option<usize>,
    /// Radius around BND sites used when building the database.
    #[arg(long, default_value_t = 50)]
    pub slack_bnd: i32,
    /// Radius around INS sites used when building the database.
    #[arg(long, default_value_t = 50)]
    pub slack_ins: i32,
    /// Minimal reciprocal overlap for SVs of the same type, used when building
    /// the database.
    #[arg(long, default_value_t = 0.8)]
    pub min_overlap: f32,
    /// Maximal distance to TAD to consider.
    #[arg(long, default_value_t = 10_000)]
    pub max_tad_distance: i32,
    /// Optional seed for RNG.
    #[arg(long)]
    pub rng_seed: Option<u64>,
}

/// Main entry point for `seqvars query` sub command.
pub fn run(args_common: &crate::common::Args, args: &Args) -> Result<(), anyhow::Error> {
    let before_anything = Instant::now();
    tracing::info!("args_common = {:?}", &args_common);
    tracing::info!("args = {:?}", &args);

    // Initialize the random number generator from command line seed if given or local entropy
    // source.
    let _rng = if let Some(rng_seed) = args.rng_seed {
        rand::rngs::StdRng::seed_from_u64(rng_seed)
    } else {
        rand::rngs::StdRng::from_entropy()
    };

    // tracing::info!("Loading query...");
    // let query: CaseQuery = serde_json::from_reader(File::open(&args.path_query_json)?)?;
    // tracing::info!(
    //     "... done loading query = {}",
    //     &serde_json::to_string(&query)?
    // );

    // tracing::info!("Loading worker databases...");
    // let before_loading = Instant::now();
    // let path_worker_db = format!("{}/worker", &args.path_db);
    // let dbs = load_databases(&path_worker_db, args.genome_release, args.max_tad_distance)?;
    // tracing::info!(
    //     "...done loading databases in {:?}",
    //     before_loading.elapsed()
    // );

    trace_rss_now();

    // tracing::info!("Loading mehari tx database...");
    // let before_loading = Instant::now();
    // let path_mehari_tx_db = format!(
    //     "{}/mehari/{}/txs.bin.zst",
    //     &args.path_db,
    //     &args.genome_release.to_string()
    // );
    // tracing::debug!("  path = {}", &path_mehari_tx_db);
    // let mehari_tx_db = mehari::annotate::seqvars::load_tx_db(&path_mehari_tx_db)?;
    // tracing::info!(
    //     "...done loading mehari tx database in {:?}",
    //     before_loading.elapsed()
    // );
    // tracing::info!("Building mehari index data structures...");
    // let before_building = Instant::now();
    // let mehari_tx_idx = TxIntervalTrees::new(&mehari_tx_db, args.genome_release.into());
    // let chrom_to_acc = ASSEMBLY_INFOS[args.genome_release.into()]
    //     .sequences
    //     .iter()
    //     .map(|record| {
    //         (
    //             annonars::common::cli::canonicalize(&record.name),
    //             record.refseq_ac.clone(),
    //         )
    //     })
    //     .collect::<HashMap<_, _>>();
    // tracing::info!(
    //     "...done building mehari index data structures in {:?}",
    //     before_building.elapsed()
    // );

    // trace_rss_now();

    // tracing::info!("Translating gene allow list...");
    // let hgvs_allowlist = if let Some(gene_allowlist) = &query.gene_allowlist {
    //     if gene_allowlist.is_empty() {
    //         None
    //     } else {
    //         Some(translate_gene_allowlist(gene_allowlist, &dbs))
    //     }
    // } else {
    //     None
    // };

    // tracing::info!("Running queries...");
    // let before_query = Instant::now();
    // let query_stats = run_query(
    //     &QueryInterpreter::new(query, hgvs_allowlist),
    //     args,
    //     &dbs,
    //     &mehari_tx_db,
    //     &mehari_tx_idx,
    //     &chrom_to_acc,
    //     &mut rng,
    // )?;
    // tracing::info!("... done running query in {:?}", before_query.elapsed());
    // tracing::info!(
    //     "summary: {} records passed out of {}",
    //     query_stats.count_passed.separate_with_commas(),
    //     query_stats.count_total.separate_with_commas()
    // );
    // tracing::info!("passing records by SV type");
    // for (sv_type, count) in query_stats.by_sv_type.iter() {
    //     tracing::info!("{:?} -- {}", sv_type, count);
    // }

    trace_rss_now();

    tracing::info!(
        "All of `seqvars query` completed in {:?}",
        before_anything.elapsed()
    );
    Ok(())
}
