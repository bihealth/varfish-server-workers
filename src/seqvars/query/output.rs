//! Data structureds for writing the output.

/// Gene-related information for the gene.
pub mod gene_related {
    use mehari::annotate::seqvars::ann::Consequence;

    use crate::seqvars::query::schema::SequenceVariant;

    /// Gene-related information for a `ResultPayload`.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Record {
        /// Gene identity related (for display of gene symbol).
        pub identity: Identity,
        /// Gene-related consequences of a variant.
        pub consequences: Consequences,
        /// Phenotype-related information, if any.
        pub phenotype: Option<Phenotype>,
        /// Gene-wise constraints on the gene, if any.
        pub constraints: Option<Constraints>,
    }

    impl Record {
        /// Construct given a `SequenceVariant` if the information is given in the annotation.
        ///
        /// Note that we will only look at the first annotation record as the ingest creates
        /// one `SequenceVariant` record per gene.
        ///
        /// # Error
        ///
        /// Returns an error if `seqvar` does not contain all necessary information.
        pub fn with_seqvar(seqvar: &SequenceVariant) -> Result<Option<Self>, anyhow::Error> {
            if let Some(ann) = seqvar.ann_fields.first() {
                if !ann.gene_id.is_empty() && !ann.gene_symbol.is_empty() {
                    return Ok(Some(Self {
                        identity: Identity::new(ann.gene_id.clone(), ann.gene_symbol.clone()),
                        consequences: Consequences::new(
                            ann.hgvs_t
                                .clone()
                                .ok_or_else(|| anyhow::anyhow!("missing hgvs_t annotation"))?,
                            ann.hgvs_p.clone(),
                            ann.consequences.clone(),
                        ),
                        // TODO: phenotype
                        // TODO: constraints
                        ..Default::default()
                    }));
                }
            }
            Ok(None)
        }
    }

    /// Result information for gene identity.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Identity {
        /// HGNC gene ID.
        pub hgnc_id: String,
        /// HGNC gene symbol.
        pub hgnc_symbol: String,
    }

    /// Result information related to phenotype / disease.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Phenotype {
        /// Whether the gene is a known disease gene.
        pub is_disease_gene: bool,
        // TODO: modes of inheritance
        // TODO: disease/phenotype terms?
    }

    /// Consequences related to a gene.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Consequences {
        /// HGVS.{c,n} code of variant
        pub hgvs_t: String,
        /// HGVS.p code of variant
        pub hgvs_p: Option<String>,

        /// The predicted variant consequences.
        pub consequences: Vec<Consequence>,
    }

    /// Result gene constraint information.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Constraints {
        /// gnomAD loeuf score
        pub gnomad_loeuf: f32,
        /// gnomAD mis_z score
        pub gnomad_mis_z: f32,
        /// gnomAD oe_lof score
        pub gnomad_oe_lof: f32,
        /// gnomAD oe_lof_lower score
        pub gnomad_oe_lof_lower: f32,
        /// gnomAD oe_lof_upper score
        pub gnomad_oe_lof_upper: f32,
        /// gnomAD oe_mis score
        pub gnomad_oe_mis: f32,
        /// gnomAD oe_mis_lower score
        pub gnomad_oe_mis_lower: f32,
        /// gnomAD oe_mis_upper score
        pub gnomad_oe_mis_upper: f32,
        /// gnomAD pLI score
        pub gnomad_pli: f32,
        /// gnomAD syn_z score
        pub gnomad_syn_z: f32,
    }
}

/// Variant-related information.
pub mod variant_related {
    use crate::seqvars::query::{annonars::Annotator, schema::SequenceVariant};

    /// Record for variant-related scores.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Record {
        /// Precomputed scores.
        pub precomputed_scores: indexmap::IndexMap<String, f32>,
        /// Database identifiers.
        pub db_ids: DbIds,
        /// Clinvar information.
        pub clinvar: Option<Clinvar>,
        /// Frequency information.
        pub frequency: Frequency,
    }

    impl Record {
        /// Construct given sequence variant and annonars annotator.
        pub fn with_seqvar_and_annotator(
            seqvar: &SequenceVariant,
            annotator: &Annotator,
        ) -> Result<Self, anyhow::Error> {
            Ok(Self {
                precomputed_scores: Self::query_precomputed_scores(seqvar, annotator)?,
                db_ids: DbIds::with_seqvar_and_annotator(seqvar, annotator)?,
                clinvar: Clinvar::with_seqvar_and_annotator(seqvar, annotator)?,
                frequency: Frequency::with_seqvar(seqvar)?,
            })
        }

        /// Query precomputed scores for `seqvar` from annonars `annotator`.
        pub fn query_precomputed_scores(
            _seqvar: &SequenceVariant,
            _annotator: &Annotator,
        ) -> Result<indexmap::IndexMap<String, f32>, anyhow::Error> {
            // TODO: implement me!
            Ok(indexmap::IndexMap::default())
        }
    }

    /// Database identifiers.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct DbIds {
        /// The variant's dbSNP identifier present.
        pub dbsnp_rs: Option<String>,
    }

    impl DbIds {
        /// Construct given sequence variant and annonars annotator.
        pub fn with_seqvar_and_annotator(
            seqvar: &SequenceVariant,
            annotator: &Annotator,
        ) -> Result<Self, anyhow::Error> {
            // TODO: need to properly separate error from no result in annonars.
            Ok(Self {
                dbsnp_rs: annotator
                    .query_dbsnp(seqvar)
                    .ok()
                    .map(|record| format!("rs{}", record.rs_id)),
            })
        }
    }

    /// ClinVar-related information.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Clinvar {
        // TODO: switch to VCV summary or hierarchical?
        /// The RCV accession.
        pub rcv: String,
        /// The clinical significance.
        pub significance: String,
        /// The review status.
        pub review_status: String,
    }

    impl Clinvar {
        /// Construct given sequence variant and annonars annotator.
        pub fn with_seqvar_and_annotator(
            seqvar: &SequenceVariant,
            annotator: &Annotator,
        ) -> Result<Option<Self>, anyhow::Error> {
            // TODO: need to properly separate error from no result in annonars.
            annotator
                .query_clinvar_minimal(seqvar)
                .ok()
                .map(|record| {
                    let annonars::clinvar_minimal::pbs::Record {
                        rcv,
                        clinical_significance,
                        review_status,
                        ..
                    } = record;
                    Ok(Self {
                        rcv,
                        significance: match clinical_significance {
                            0 => "Pathogenic",
                            1 => "Likely pathogenic",
                            2 => "Uncertain significance",
                            3 => "Likely benign",
                            4 => "Benign",
                            _ => anyhow::bail!(
                                "invalid clinical significance enum: {}",
                                clinical_significance
                            ),
                        }
                        .into(),
                        review_status: match review_status {
                            0 => "no assertion provided",
                            1 => "no assertion criteria provided",
                            2 => "criteria provided, conflicting interpretations",
                            3 => "criteria provided, single submitter",
                            4 => "criteria provided, multiple submitters, no conflicts",
                            5 => "reviewed by expert panel",
                            6 => "practice guideline",
                            _ => anyhow::bail!("invalid review status enum: {}", review_status),
                        }
                        .into(),
                    })
                })
                .transpose()
        }
    }

    /// Frequency information.
    #[derive(
        Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_builder::Builder,
    )]
    pub struct Frequency {
        /// gnomAD-genomes frequency
        pub gnomad_genomes: Option<NuclearFrequency>,
        /// gnomAD-exomes frequency
        pub gnomad_exomes: Option<NuclearFrequency>,
        /// gnomad-mtDNA frequency
        pub gnomad_mtdna: Option<MtdnaFrequency>,
        /// HelixMtDb frequency
        pub helixmtdb: Option<MtdnaFrequency>,
        /// in-house frequency
        pub inhouse: Option<NuclearFrequency>,
    }

    impl Frequency {
        /// Extract frequency information from `seqvar`
        pub fn with_seqvar(seqvar: &SequenceVariant) -> Result<Frequency, anyhow::Error> {
            let chrom = annonars::common::cli::canonicalize(&seqvar.chrom);
            let frequency = if chrom == "MT" {
                FrequencyBuilder::default()
                    .gnomad_genomes(Some(NuclearFrequency::new(
                        seqvar.gnomad_genomes_af(),
                        seqvar.gnomad_genomes_an,
                        seqvar.gnomad_genomes_het,
                        seqvar.gnomad_genomes_hom,
                        seqvar.gnomad_genomes_hemi,
                    )))
                    .gnomad_exomes(Some(NuclearFrequency::new(
                        seqvar.gnomad_exomes_af(),
                        seqvar.gnomad_exomes_an,
                        seqvar.gnomad_exomes_het,
                        seqvar.gnomad_exomes_hom,
                        seqvar.gnomad_exomes_hemi,
                    )))
                    .build()
            } else {
                FrequencyBuilder::default()
                    .gnomad_mtdna(Some(MtdnaFrequency::new(
                        seqvar.gnomad_genomes_af(),
                        seqvar.gnomad_genomes_an,
                        seqvar.gnomad_genomes_het,
                        seqvar.gnomad_genomes_hom,
                    )))
                    .helixmtdb(Some(MtdnaFrequency::new(
                        seqvar.helixmtdb_af(),
                        seqvar.helix_an,
                        seqvar.helix_het,
                        seqvar.helix_hom,
                    )))
                    .build()
            }
            .map_err(|e| anyhow::anyhow!("could not build frequency information: {}", e))?;
            Ok(frequency)
        }
    }

    /// Nuclear frequency information.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct NuclearFrequency {
        /// Overall allele frequency.
        pub allele_freq: f32,
        /// Number of alleles.
        pub allele_count: i32,
        /// Number of heterozygous carriers.
        pub het_carriers: i32,
        /// Number of homozygous carriers.
        pub hom_carriers: i32,
        /// Number of hemizygous carriers.
        pub hemi_carriers: i32,
    }

    /// Mitochondrial frequency information.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct MtdnaFrequency {
        /// Overall allele frequency.
        pub allele_freq: f32,
        /// Number of alleles.
        pub allele_count: i32,
        /// Number of heterplasmic carriers.
        pub het_carriers: i32,
        /// Number of homoplasmic carriers.
        pub hom_carriers: i32,
    }
}

/// Call-related information.
pub mod call_related {
    use crate::seqvars::query::schema::SequenceVariant;

    /// Call-related record.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct Record {
        /// The genotype information for each sample.
        pub call_info: indexmap::IndexMap<String, CallInfo>,
    }

    impl Record {
        /// Construct a new `Record` from a `SequenceVariant`.
        ///
        /// # Error
        ///
        /// Returns an error if the `SequenceVariant` does not contain all necessary information.
        pub fn with_seqvar(seqvar: &SequenceVariant) -> Result<Self, anyhow::Error> {
            Ok(Self {
                call_info: seqvar
                    .call_info
                    .iter()
                    .map(|(sample_name, call_info)| {
                        (
                            sample_name.clone(),
                            CallInfo::new(
                                call_info.dp,
                                call_info.ad,
                                call_info.quality.map(|q| q as i32),
                                call_info.genotype.clone(),
                            ),
                        )
                    })
                    .collect(),
            })
        }
    }

    /// Genotype information.
    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_new::new)]
    pub struct CallInfo {
        /// Depth of coverage.
        pub dp: Option<i32>,
        /// Alternate read depth.
        pub ad: Option<i32>,
        /// Genotype quality.
        pub gq: Option<i32>,
        /// Genotype.
        pub gt: Option<String>,
    }
}

/// Result record information.
pub mod result {
    /// A result record from the query.
    ///
    /// These records are written to TSV for import into the database.   They contain the
    /// bare necessary information for sorting etc. in the database.  The actual main
    /// data is in the payload, serialized as JSON.
    #[derive(
        Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_builder::Builder,
    )]
    pub struct Record {
        /// UUID for the record.
        pub sodar_uuid: uuid::Uuid,
        /// Genome release for the coordinate.
        pub release: String,
        /// Chromosome name.
        pub chromosome: String,
        /// Chromosome number.
        pub chromosome_no: i32,
        /// Reference allele sequence.
        pub reference: String,
        /// Alternative allele sequence.
        pub alternative: String,
        /// UCSC bin of the record.
        pub bin: u32,
        /// Start position of the record.
        pub start: i32,
        /// End position of the record.
        pub end: i32,
        /// The result set ID as specified on the command line.
        pub smallvariantqueryresultset_id: String,
        /// The JSON-serialized `ResultPayload`.
        pub payload: String,
    }

    /// The structured result information of the result record.
    #[derive(
        Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_builder::Builder,
    )]
    pub struct Payload {
        /// Case UUID as specified on the command line.
        pub case_uuid: uuid::Uuid,
        /// The affected gene and consequence, if any.
        pub gene_related: Option<super::gene_related::Record>,
        /// Variant-related information, always present.
        pub variant_related: super::variant_related::Record,
        /// Genotypes call related, always present.
        pub call_related: super::call_related::Record,
    }
}
