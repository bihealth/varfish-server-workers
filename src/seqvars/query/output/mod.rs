//! Data structureds for writing the output.

pub mod gene_related;
pub mod variant_related;

pub mod call_related;

/// A result record from the query.
///
/// These records are written to TSV for import into the database.   They contain the
/// bare necessary information for sorting etc. in the database.  The actual main
/// data is in the payload, serialized as JSON.
#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_builder::Builder)]
pub struct Record {
    /// UUID for the record.
    pub sodar_uuid: uuid::Uuid,
    /// Genome release for the coordinate.
    pub release: String,
    /// Chromosome name.
    pub chromosome: String,
    /// Chromosome number.
    pub chromosome_no: i32,
    /// Reference allele sequence.
    pub reference: String,
    /// Alternative allele sequence.
    pub alternative: String,
    /// UCSC bin of the record.
    pub bin: u32,
    /// Start position of the record.
    pub start: i32,
    /// End position of the record.
    pub end: i32,
    /// The result set ID as specified on the command line.
    pub smallvariantqueryresultset_id: String,
    /// The JSON-serialized `ResultPayload`.
    pub payload: String,
}

/// The structured result information of the result record.
#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize, derive_builder::Builder)]
pub struct Payload {
    /// Case UUID as specified on the command line.
    pub case_uuid: uuid::Uuid,
    /// The affected gene and consequence, if any.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gene_related: Option<gene_related::Record>,
    /// Variant-related information, always present.
    #[serde(skip_serializing_if = "variant_related::Record::is_empty")]
    pub variant_related: variant_related::Record,
    /// Genotypes call related, always present.
    pub call_related: call_related::Record,
}
