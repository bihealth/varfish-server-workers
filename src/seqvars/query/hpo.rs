//! Code for accessing HPO-related information.

use crate::pbs::varfish::v1::seqvars::output as pbs_output;

/// Enumeration for modes of inheritance.
#[derive(
    serde::Serialize,
    serde::Deserialize,
    enum_map::Enum,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Clone,
    Copy,
    Debug,
    strum::EnumString,
    strum::Display,
)]
pub enum ModeOfInheritance {
    /// Autosomal dominant inheritance (HP:0000006).
    AutosomalDominant,
    /// Autosomal recessive inheritance (HP:0000007).
    AutosomalRecessive,
    /// X-linked dominant inheritance (HP:0001419).
    XLinkedDominant,
    /// X-linked recessive inheritance (HP:0001423).
    XLinkedRecessive,
    /// Y-linked inheritance (HP:0001450).
    YLinked,
    /// Mitochondrial inheritance (HP:0001427).
    Mitochondrial,
}

impl ModeOfInheritance {
    /// Allow parsing of `ModeOfInheritance` from HPO ID.
    pub fn from_hpo_id(hpo_id: &str) -> Option<Self> {
        match hpo_id {
            "HP:0000006" | "HP:0012275" => Some(Self::AutosomalDominant),
            "HP:0000007" => Some(Self::AutosomalRecessive),
            "HP:0001419" => Some(Self::XLinkedDominant),
            "HP:0001423" => Some(Self::XLinkedRecessive),
            "HP:0001450" => Some(Self::YLinked),
            "HP:0001427" => Some(Self::Mitochondrial),
            _ => None,
        }
    }
}

impl From<ModeOfInheritance> for pbs_output::ModeOfInheritance {
    fn from(val: ModeOfInheritance) -> Self {
        match val {
            ModeOfInheritance::AutosomalDominant => {
                pbs_output::ModeOfInheritance::AutosomalDominant
            }
            ModeOfInheritance::AutosomalRecessive => {
                pbs_output::ModeOfInheritance::AutosomalRecessive
            }
            ModeOfInheritance::XLinkedDominant => pbs_output::ModeOfInheritance::XLinkedDominant,
            ModeOfInheritance::XLinkedRecessive => pbs_output::ModeOfInheritance::XLinkedRecessive,
            ModeOfInheritance::YLinked => pbs_output::ModeOfInheritance::YLinked,
            ModeOfInheritance::Mitochondrial => pbs_output::ModeOfInheritance::Mitochondrial,
        }
    }
}

pub type HgncToMoiMap = indexmap::IndexMap<String, indexmap::IndexSet<ModeOfInheritance>>;

/// Load the the `phenotype_to_genes.tsv` and `hgnc_xlink.tsv` files from the `hpo`
/// directory and build a map from HGNC gene ID to set of `ModeOfInheritance`
/// values.
pub fn load_hgnc_to_inheritance_map<P: AsRef<std::path::Path>>(
    path: &P,
) -> Result<HgncToMoiMap, anyhow::Error> {
    let phenotypes_to_genes =
        phenotype_to_genes::load_entries(&path.as_ref().join("phenotype_to_genes.txt"))
            .map_err(|e| anyhow::anyhow!("error loading phenotype_to_genes.txt: {}", e))?;
    let ncbi_to_hgnc = hgnc_xlink::load_ncbi_to_hgnc(path.as_ref().join("hgnc_xlink.tsv"))
        .map_err(|e| anyhow::anyhow!("error loading hgnc_xlink.tsv: {}", e))?;

    let mut result = indexmap::IndexMap::new();

    for ptg_entry in phenotypes_to_genes {
        if let Some(ncbi_gene_id) = ptg_entry.ncbi_gene_id {
            if let Some(hgnc_ids) = ncbi_to_hgnc.get(&ncbi_gene_id) {
                for hgnc_id in hgnc_ids {
                    if let Some(mode_of_inheritance) =
                        ModeOfInheritance::from_hpo_id(&ptg_entry.hpo_id)
                    {
                        let modes_of_inheritance = result
                            .entry(hgnc_id.clone())
                            .or_insert_with(indexmap::IndexSet::new);
                        modes_of_inheritance.insert_sorted(mode_of_inheritance);
                    }
                }
            }
        }
    }

    Ok(result)
}

/// Code for accessing the `phenotype_to_genes.tsv` file.
pub(super) mod phenotype_to_genes {
    /// Data structure for representing an entry of the table.
    #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
    #[serde_with::skip_serializing_none]
    pub struct Entry {
        /// Entrez gene ID.
        #[serde(alias = "entrez_id")]
        pub ncbi_gene_id: Option<u32>,
        /// Gene symbol.
        pub gene_symbol: String,
        /// HPO ID.
        pub hpo_id: String,
        // HPO Name.
        pub hpo_name: String,
    }

    /// Read the `phenotype_to_genes.tsv` file using the `csv` crate via serde.
    ///
    /// # Errors
    ///
    /// In the case that the file could not be read.
    pub fn load_entries<P: AsRef<std::path::Path>>(path: &P) -> Result<Vec<Entry>, anyhow::Error> {
        let mut rdr = csv::ReaderBuilder::new()
            .delimiter(b'\t')
            .has_headers(true)
            .from_path(path.as_ref())?;
        let mut entries = Vec::new();
        for result in rdr.deserialize() {
            let entry: Entry = result?;
            entries.push(entry);
        }
        Ok(entries)
    }
}

/// Code for accessing the `hgnc_xlink.tsv` file.
pub(super) mod hgnc_xlink {
    use std::collections::HashMap;

    /// Data structure for representing an entry of the table.
    #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
    #[serde_with::skip_serializing_none]
    pub struct Entry {
        /// HGNC gene ID.
        pub hgnc_id: String,
        /// Ensembl gene ID.
        pub ensembl_gene_id: Option<String>,
        /// Entrez gene ID.
        #[serde(alias = "entrez_id")]
        pub ncbi_gene_id: Option<u32>,
        /// Gene symbol.
        pub gene_symbol: String,
    }

    /// Read the `hgnc_xlink.tsv` file using the `csv` crate via serde.
    ///
    /// # Errors
    ///
    /// In the case that the file could not be read.
    pub fn load_entries<P: AsRef<std::path::Path>>(path: &P) -> Result<Vec<Entry>, anyhow::Error> {
        let mut rdr = csv::ReaderBuilder::new()
            .delimiter(b'\t')
            .has_headers(true)
            .from_path(path.as_ref())?;
        let mut entries = Vec::new();
        for result in rdr.deserialize() {
            let entry: Entry = result?;
            entries.push(entry);
        }
        Ok(entries)
    }

    /// Read the `hgnc_xlink.tsv` into a map from NCBI gene ID to HGNC gene ID.
    ///
    /// # Errors
    ///
    /// In the case that the file could not be read.
    pub fn load_ncbi_to_hgnc<P: AsRef<std::path::Path>>(
        path: P,
    ) -> Result<HashMap<u32, Vec<String>>, anyhow::Error> {
        let mut map = HashMap::new();
        for entry in load_entries(&path)? {
            if let Some(ncbi_gene_id) = entry.ncbi_gene_id {
                let hgnc_id = entry.hgnc_id;
                let entry = map.entry(ncbi_gene_id).or_insert_with(Vec::new);
                entry.push(hgnc_id);
            }
        }
        Ok(map)
    }
}

#[cfg(test)]
mod test {
    #[test]
    pub fn test_mode_of_inheritance_from_hpo_id() {
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0000006"),
            Some(super::ModeOfInheritance::AutosomalDominant)
        );
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0012275"),
            Some(super::ModeOfInheritance::AutosomalDominant)
        );
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0000007"),
            Some(super::ModeOfInheritance::AutosomalRecessive)
        );
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0001419"),
            Some(super::ModeOfInheritance::XLinkedDominant)
        );
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0001423"),
            Some(super::ModeOfInheritance::XLinkedRecessive)
        );
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0001450"),
            Some(super::ModeOfInheritance::YLinked)
        );
        assert_eq!(
            super::ModeOfInheritance::from_hpo_id("HP:0001427"),
            Some(super::ModeOfInheritance::Mitochondrial)
        );

        assert_eq!(super::ModeOfInheritance::from_hpo_id("HP:0000000"), None);
    }

    #[test]
    pub fn test_mode_of_inheritance_into_pbs() {
        assert_eq!(
            Into::<super::pbs_output::ModeOfInheritance>::into(
                super::ModeOfInheritance::AutosomalDominant
            ),
            super::pbs_output::ModeOfInheritance::AutosomalDominant
        );
        assert_eq!(
            Into::<super::pbs_output::ModeOfInheritance>::into(
                super::ModeOfInheritance::AutosomalRecessive
            ),
            super::pbs_output::ModeOfInheritance::AutosomalRecessive
        );
        assert_eq!(
            Into::<super::pbs_output::ModeOfInheritance>::into(
                super::ModeOfInheritance::XLinkedDominant
            ),
            super::pbs_output::ModeOfInheritance::XLinkedDominant
        );
        assert_eq!(
            Into::<super::pbs_output::ModeOfInheritance>::into(
                super::ModeOfInheritance::XLinkedRecessive
            ),
            super::pbs_output::ModeOfInheritance::XLinkedRecessive
        );
        assert_eq!(
            Into::<super::pbs_output::ModeOfInheritance>::into(super::ModeOfInheritance::YLinked),
            super::pbs_output::ModeOfInheritance::YLinked
        );
        assert_eq!(
            Into::<super::pbs_output::ModeOfInheritance>::into(
                super::ModeOfInheritance::Mitochondrial
            ),
            super::pbs_output::ModeOfInheritance::Mitochondrial
        );
    }

    #[test]
    fn load_hgnc_to_inheritance_map() -> Result<(), anyhow::Error> {
        let path = std::path::Path::new("tests/seqvars/query/db/hpo");
        let map = super::load_hgnc_to_inheritance_map(&path)?;

        assert_eq!(map.len(), 4599);
        insta::assert_yaml_snapshot!(&map[0..5]);

        Ok(())
    }

    #[test]
    fn test_phenotype_to_genes_load_entries() -> Result<(), anyhow::Error> {
        let path = std::path::Path::new("tests/seqvars/query/db/hpo/phenotype_to_genes.txt");
        let entries = super::phenotype_to_genes::load_entries(&path)?;

        assert_eq!(entries.len(), 988720);
        insta::assert_yaml_snapshot!(&entries[0..5]);

        Ok(())
    }

    #[test]
    fn test_hgnc_to_xlink_load_entries() -> Result<(), anyhow::Error> {
        let path = std::path::Path::new("tests/seqvars/query/db/hpo/hgnc_xlink.tsv");
        let map = super::hgnc_xlink::load_entries(&path)?;

        assert_eq!(map.len(), 43839);
        insta::assert_yaml_snapshot!(&map[0..5]);

        Ok(())
    }

    #[test]
    fn test_load_ncbi_to_hgnc() -> Result<(), anyhow::Error> {
        let path = std::path::Path::new("tests/seqvars/query/db/hpo/hgnc_xlink.tsv");
        let map = super::hgnc_xlink::load_ncbi_to_hgnc(&path)?;

        assert_eq!(map.len(), 43792);
        assert_eq!(map.get(&1), Some(&vec!["HGNC:5".to_string()]));
        assert_eq!(map.get(&2), Some(&vec!["HGNC:7".to_string()]));

        Ok(())
    }
}
