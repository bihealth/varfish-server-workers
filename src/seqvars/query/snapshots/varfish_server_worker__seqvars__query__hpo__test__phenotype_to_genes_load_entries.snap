---
source: src/seqvars/query/hpo.rs
expression: "&entries[0..5]"
---
- ncbi_gene_id: 3081
  gene_symbol: HGD
  hpo_id: "HP:0008775"
  hpo_name: Abnormal prostate morphology
- ncbi_gene_id: 695
  gene_symbol: BTK
  hpo_id: "HP:0008775"
  hpo_name: Abnormal prostate morphology
- ncbi_gene_id: 695
  gene_symbol: BTK
  hpo_id: "HP:0008775"
  hpo_name: Abnormal prostate morphology
- ncbi_gene_id: 1493
  gene_symbol: CTLA4
  hpo_id: "HP:0008775"
  hpo_name: Abnormal prostate morphology
- ncbi_gene_id: 26191
  gene_symbol: PTPN22
  hpo_id: "HP:0008775"
  hpo_name: Abnormal prostate morphology
